When receiving events from the client, Flow should not throw an error if the event data is null

Some components fire events without properties when those properties are `null` (or `undefined`). For example, `vaadin-combo-box` fires the `selected-item-changed` event without the `event.detail.value` property when there's no selected value.

That causes a server-wide error which results in a scary message at the client side and server instability after it:

> SEVERE: 
java.lang.IllegalArgumentException: Unable to create an event object of type com.vaadin.generated.vaadin.combo.box.GeneratedVaadinComboBox$SelectedItemChangeEvent
	at com.vaadin.flow.event.ComponentEventBus.createEventForDomEvent(ComponentEventBus.java:335)
	at com.vaadin.flow.event.ComponentEventBus.handleDomEvent(ComponentEventBus.java:291)
	at com.vaadin.flow.event.ComponentEventBus.access$200(ComponentEventBus.java:49)
	at com.vaadin.flow.event.ComponentEventBus$1.handleEvent(ComponentEventBus.java:180)
	at com.vaadin.flow.nodefeature.ElementListenerMap.lambda$fireEvent$0(ElementListenerMap.java:169)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.vaadin.flow.nodefeature.ElementListenerMap.fireEvent(ElementListenerMap.java:169)
	at com.vaadin.server.communication.rpc.EventRpcHandler.handleNode(EventRpcHandler.java:56)
	at com.vaadin.server.communication.rpc.AbstractRpcInvocationHandler.handle(AbstractRpcInvocationHandler.java:50)
	at com.vaadin.server.communication.ServerRpcHandler.handleInvocationData(ServerRpcHandler.java:393)
	at com.vaadin.server.communication.ServerRpcHandler.lambda$handleInvocations$0(ServerRpcHandler.java:383)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.vaadin.server.communication.ServerRpcHandler.handleInvocations(ServerRpcHandler.java:383)
	at com.vaadin.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:303)
	at com.vaadin.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:87)
	at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1417)
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:294)
[...]
Caused by: java.lang.IllegalArgumentException: The DOM event does not contain the expected event data: event.detail.value
	at com.vaadin.flow.event.ComponentEventBus.lambda$createEventDataObjects$4(ComponentEventBus.java:207)
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at com.vaadin.flow.event.ComponentEventBus.createEventDataObjects(ComponentEventBus.java:204)
	at com.vaadin.flow.event.ComponentEventBus.createEventForDomEvent(ComponentEventBus.java:321)

Reference code: https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/event/ComponentEventBus.java#L207

Flow should be able to create the event even without the properties inside it - just assume the properties are `null`.

--- a/flow-server/src/main/java/com/vaadin/flow/event/ComponentEventBus.java
+++ b/flow-server/src/main/java/com/vaadin/flow/event/ComponentEventBus.java
@@ -31,6 +31,7 @@ import com.vaadin.shared.Registration;
 import com.vaadin.ui.Component;
 import com.vaadin.ui.ComponentEvent;
 
+import elemental.json.Json;
 import elemental.json.JsonValue;
 
 /**
@@ -139,7 +140,7 @@ public class ComponentEventBus implements Serializable {
         }
 
         ComponentEventBusUtil.getDomEventType(eventType)
-                .ifPresent(e -> addDomTrigger(eventType, e));
+        .ifPresent(e -> addDomTrigger(eventType, e));
     }
 
     /**
@@ -156,7 +157,7 @@ public class ComponentEventBus implements Serializable {
             String domEventType) {
         assert eventType != null;
         assert !componentEventData.containsKey(eventType)
-                || componentEventData.get(eventType).domEventRemover == null;
+        || componentEventData.get(eventType).domEventRemover == null;
 
         if (domEventType == null || domEventType.isEmpty()) {
             throw new IllegalArgumentException(
@@ -175,11 +176,11 @@ public class ComponentEventBus implements Serializable {
         // https://github.com/vaadin/flow/issues/575
         Registration remover = element.addEventListener(domEventType,
                 new DomEventListener() {
-                    @Override
-                    public void handleEvent(DomEvent e) {
-                        handleDomEvent(eventType, e);
-                    }
-                }, eventData);
+            @Override
+            public void handleEvent(DomEvent e) {
+                handleDomEvent(eventType, e);
+            }
+        }, eventData);
         componentEventData.computeIfAbsent(eventType,
                 t -> new ComponentEventData()).domEventRemover = remover;
     }
@@ -204,9 +205,7 @@ public class ComponentEventBus implements Serializable {
         expressions.forEach((expression, type) -> {
             JsonValue jsonValue = domEvent.getEventData().get(expression);
             if (jsonValue == null) {
-                throw new IllegalArgumentException(
-                        "The DOM event does not contain the expected event data: "
-                                + expression);
+                jsonValue = Json.createNull();
             }
             Object value = JsonCodec.decodeAs(jsonValue, type);
             eventDataObjects.add(value);
@@ -245,7 +244,7 @@ public class ComponentEventBus implements Serializable {
         if (listeners.isEmpty()) {
             // No more listeners for this event type
             ComponentEventBusUtil.getDomEventType(eventType)
-                    .ifPresent(e -> unregisterDomEvent(eventType, e));
+            .ifPresent(e -> unregisterDomEvent(eventType, e));
             componentEventData.remove(eventType);
         }
     }
@@ -333,7 +332,7 @@ public class ComponentEventBus implements Serializable {
             throw new IllegalArgumentException(
                     "Unable to create an event object of type "
                             + eventType.getName(),
-                    e);
+                            e);
         }
     }
 }

--- a/flow-server/src/test/java/com/vaadin/flow/event/ComponentEventBusTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/event/ComponentEventBusTest.java
@@ -36,7 +36,7 @@ import elemental.json.JsonObject;
 public class ComponentEventBusTest {
 
     private static class EventTracker<T extends ComponentEvent<?>>
-            implements ComponentEventListener<T> {
+    implements ComponentEventListener<T> {
         private AtomicInteger eventHandlerCalled = new AtomicInteger(0);
         private AtomicReference<T> eventObject = new AtomicReference<>(null);
 
@@ -92,8 +92,8 @@ public class ComponentEventBusTest {
             JsonObject eventData) {
         Element e = component.getElement();
         e.getNode().getFeature(ElementListenerMap.class)
-                .fireEvent(new com.vaadin.flow.dom.DomEvent(e, domEvent,
-                        eventData));
+        .fireEvent(new com.vaadin.flow.dom.DomEvent(e, domEvent,
+                eventData));
 
     }
 
@@ -112,7 +112,7 @@ public class ComponentEventBusTest {
 
         eventTracker.reset();
         component.getEventBus()
-                .fireEvent(new MappedToDomEvent(component, true));
+        .fireEvent(new MappedToDomEvent(component, true));
 
         eventTracker.assertEventCalled(component, true);
         Assert.assertEquals(12, eventTracker.getEvent().getSomeData());
@@ -133,7 +133,7 @@ public class ComponentEventBusTest {
         });
 
         component
-                .fireEvent(new ServerEvent(component, new BigDecimal("12.22")));
+        .fireEvent(new ServerEvent(component, new BigDecimal("12.22")));
 
         Assert.assertEquals(1, eventHandlerCalled.get());
         Assert.assertEquals(new BigDecimal("12.22"), dataValueInEvent.get());
@@ -161,12 +161,15 @@ public class ComponentEventBusTest {
         fireDomEvent(c, "dom-event", Json.createObject());
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    @Test
     public void mappedDomEvent_fire_missingData() {
         TestComponent c = new TestComponent();
         EventTracker<MappedToDomEvent> eventListener = new EventTracker<>();
         c.addListener(MappedToDomEvent.class, eventListener);
         fireDomEvent(c, "dom-event", createData("event.someData", 2));
+        eventListener.assertEventCalled(c, true);
+        Assert.assertEquals(2, eventListener.getEvent().getSomeData());
+        Assert.assertNull(eventListener.getEvent().getMoreData());
     }
 
     private JsonObject createData(String key, Object value) {

